File: Python/minNumber.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: minNumber.py
@Function: 把数组排成最小的数
@Link: https://leetcode-cn.com/problems/ba-shu-zu-pai-cheng-zui-xiao-de-shu-lcof/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-09-17
"""


from typing import List


class Solution:
    def minNumber(self, nums: List[int]) -> str:
        def quick_sort(l, r):
            if l >= r:
                return
            i, j = l, r
            while i < j:
                while strs[j] + strs[l] >= strs[l] + strs[j] and i < j:
                    j -= 1
                while strs[i] + strs[l] <= strs[l] + strs[i] and i < j:
                    i += 1
                strs[i], strs[j] = strs[j], strs[i]
            strs[i], strs[l] = strs[l], strs[i]
            quick_sort(l, i - 1)
            quick_sort(i + 1, r)

        strs = [str(num) for num in nums]
        quick_sort(0, len(strs) - 1)
        return ''.join(strs)


# -------------------------
if __name__ == "__main__":
    # nums = [3,30,34,5,9]
    nums = [10, 2]

    solution = Solution()
    min_number = solution.minNumber(nums)
    print(f"The solution of this problem is {min_number}")

File: C++/minNumber.cpp
/**
 * @File    : minNumber.cpp
 * @Brief   : 把数组排成最小的数
 * @Link    : https://leetcode-cn.com/problems/ba-shu-zu-pai-cheng-zui-xiao-de-shu-lcof/
 * @Author  : Wei Li
 * @Date    : 2021-09-17
*/

#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

class Solution
{
public:
    std::string minNumber(std::vector<int> &nums)
    {
        std::vector<std::string> strs;
        for (int i = 0; i < nums.size(); i++)
            strs.push_back(std::to_string(nums[i]));
        quickSort(strs, 0, strs.size() - 1);
        std::string res;
        for (std::string s : strs)
            res.append(s);
        return res;
    }

private:
    void quickSort(std::vector<std::string> &strs, int l, int r)
    {
        if (l >= r)
            return;
        int i = l, j = r;
        while (i < j)
        {
            while (strs[j] + strs[l] >= strs[l] + strs[j] && i < j)
                j--;
            while (strs[i] + strs[l] <= strs[l] + strs[i] && i < j)
                i++;
            std::swap(strs[i], strs[j]);
        }
        std::swap(strs[i], strs[l]);
        quickSort(strs, l, i - 1);
        quickSort(strs, i + 1, r);
    }
};

// -----------------------------
int main(int argc, char **argv)
{
    // std::vector<int> nums{10, 2};
    std::vector<int> nums{3, 30, 34, 5, 9};

    Solution solution;
    auto min_number = solution.minNumber(nums);
    std::cout << "The solution of this problem is " << min_number << std::endl;

    return 0;
}
